// manseryeok/dist/constants.js
"use strict";
/**
 * 만세력 계산에 사용되는 상수 정의
 * Constants for Manseryeok / Saju calculation
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DAY_PILLAR_ANCHOR = exports.TEN_GOD_HANJA = exports.ELEMENT_CONTROLS = exports.ELEMENT_GENERATES = exports.MONTH_BRANCHES = exports.BRANCH_MAIN_STEM = exports.BRANCH_ELEMENTS = exports.STEM_ELEMENTS = exports.FIVE_ELEMENTS = exports.YIN_YANG = exports.EARTHLY_BRANCHES_HANJA = exports.EARTHLY_BRANCHES = exports.HEAVENLY_STEMS_HANJA = exports.HEAVENLY_STEMS = void 0;
// ===== 천간 (Heavenly Stems) =====
/** 천간 (한글) */
exports.HEAVENLY_STEMS = ['갑', '을', '병', '정', '무', '기', '경', '신', '임', '계'];
/** 천간 (한자) */
exports.HEAVENLY_STEMS_HANJA = [
    '甲',
    '乙',
    '丙',
    '丁',
    '戊',
    '己',
    '庚',
    '辛',
    '壬',
    '癸',
];
// ===== 지지 (Earthly Branches) =====
/** 지지 (한글) */
exports.EARTHLY_BRANCHES = [
    '자',
    '축',
    '인',
    '묘',
    '진',
    '사',
    '오',
    '미',
    '신',
    '유',
    '술',
    '해',
];
/** 지지 (한자) */
exports.EARTHLY_BRANCHES_HANJA = [
    '子',
    '丑',
    '寅',
    '卯',
    '辰',
    '巳',
    '午',
    '未',
    '申',
    '酉',
    '戌',
    '亥',
];
// ===== 음양오행 =====
/** 음양 (Yin/Yang) */
exports.YIN_YANG = ['양', '음'];
/** 오행 (Five Elements) */
exports.FIVE_ELEMENTS = ['목', '화', '토', '금', '수'];
/** 천간의 오행 (인덱스 기준: 갑을=목, 병정=화, 무기=토, 경신=금, 임계=수) */
exports.STEM_ELEMENTS = [
    '목',
    '목',
    '화',
    '화',
    '토',
    '토',
    '금',
    '금',
    '수',
    '수',
];
/** 지지의 오행 (인덱스 기준) */
exports.BRANCH_ELEMENTS = [
    '수', // 자
    '토', // 축
    '목', // 인
    '목', // 묘
    '토', // 진
    '화', // 사
    '화', // 오
    '토', // 미
    '금', // 신
    '금', // 유
    '토', // 술
    '수', // 해
];
// ===== 지장간 본기 =====
/** 지지의 본기(정기) 천간 */
exports.BRANCH_MAIN_STEM = {
    자: '계',
    축: '기',
    인: '갑',
    묘: '을',
    진: '무',
    사: '병',
    오: '정',
    미: '기',
    신: '경',
    유: '신',
    술: '무',
    해: '임',
};
// ===== 월지 매핑 (절기 월 → 지지) =====
/**
 * 절기 기준 월(1=인월 … 12=축월)에 대응하는 지지.
 * 인월(1)부터 시작하여 자월(11), 축월(12)로 끝난다.
 */
exports.MONTH_BRANCHES = {
    1: '인',
    2: '묘',
    3: '진',
    4: '사',
    5: '오',
    6: '미',
    7: '신',
    8: '유',
    9: '술',
    10: '해',
    11: '자',
    12: '축',
};
// ===== 오행 상생·상극 =====
/** 오행 상생: A 가 생하는 오행 (목→화→토→금→수→목) */
exports.ELEMENT_GENERATES = {
    목: '화',
    화: '토',
    토: '금',
    금: '수',
    수: '목',
};
/** 오행 상극: A 가 극하는 오행 (목→토→수→화→금→목) */
exports.ELEMENT_CONTROLS = {
    목: '토',
    토: '수',
    수: '화',
    화: '금',
    금: '목',
};
// ===== 십신 (Ten Gods) =====
/** 십신 한자 표기 */
exports.TEN_GOD_HANJA = {
    비견: '比肩',
    겁재: '劫財',
    식신: '食神',
    상관: '傷官',
    편재: '偏財',
    정재: '正財',
    편관: '偏官',
    정관: '正官',
    편인: '偏印',
    정인: '正印',
};
// ===== 일주 기준일 (60갑자 anchor) =====
/**
 * 일주 계산 기준일: 1992-10-24 (KST) = 계유일.
 * 계유 = 60갑자에서 (천간 계=9, 지지 유=9) → 인덱스 9.
 */
exports.DAY_PILLAR_ANCHOR = {
    year: 1992,
    month: 10,
    day: 24,
    ganjiIndex: 9,
};
