// manseryeok/dist/index.js
"use strict";
/**
 * 만세력(萬歲曆) 계산 라이브러리
 * Korean Saju (Four Pillars) and Manseryeok calculation library
 *
 * @author Yoohyojun
 * @license MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_LONGITUDE = exports.getLuckPillars = exports.getVoidBranches = exports.getTenGodChart = exports.getBranchTenGod = exports.getTenGod = exports.equationOfTimeMinutes = exports.apparentSolarLongitude = exports.SOLAR_TERM_NAMES_HANJA = exports.SOLAR_TERM_NAMES = exports.getSolarTermsOfYear = exports.getSolarTerm = exports.LUNAR_MAX_YEAR = exports.LUNAR_MIN_YEAR = exports.isValidSolarDate = exports.solarToLunar = exports.lunarToSolar = exports.getEarthlyBranchElement = exports.getEarthlyBranchYinYang = exports.getHeavenlyStemElement = exports.getHeavenlyStemYinYang = exports.TEN_GOD_HANJA = exports.FIVE_ELEMENTS = exports.YIN_YANG = exports.EARTHLY_BRANCHES_HANJA = exports.EARTHLY_BRANCHES = exports.HEAVENLY_STEMS_HANJA = exports.HEAVENLY_STEMS = void 0;
exports.calculateFourPillars = calculateFourPillars;
exports.fourPillarsToString = fourPillarsToString;
const constants_1 = require("./constants");
const elements_1 = require("./elements");
const convert_1 = require("./calendar/convert");
const lunar_data_1 = require("./calendar/lunar-data");
const solar_terms_data_1 = require("./astro/solar-terms-data");
const true_solar_time_1 = require("./time/true-solar-time");
const pillars_1 = require("./pillars");
const ten_gods_1 = require("./features/ten-gods");
const void_branches_1 = require("./features/void-branches");
const luck_pillars_1 = require("./features/luck-pillars");
const validation_1 = require("./validation");
// ===== 공개 re-export =====
var constants_2 = require("./constants");
Object.defineProperty(exports, "HEAVENLY_STEMS", { enumerable: true, get: function () { return constants_2.HEAVENLY_STEMS; } });
Object.defineProperty(exports, "HEAVENLY_STEMS_HANJA", { enumerable: true, get: function () { return constants_2.HEAVENLY_STEMS_HANJA; } });
Object.defineProperty(exports, "EARTHLY_BRANCHES", { enumerable: true, get: function () { return constants_2.EARTHLY_BRANCHES; } });
Object.defineProperty(exports, "EARTHLY_BRANCHES_HANJA", { enumerable: true, get: function () { return constants_2.EARTHLY_BRANCHES_HANJA; } });
Object.defineProperty(exports, "YIN_YANG", { enumerable: true, get: function () { return constants_2.YIN_YANG; } });
Object.defineProperty(exports, "FIVE_ELEMENTS", { enumerable: true, get: function () { return constants_2.FIVE_ELEMENTS; } });
Object.defineProperty(exports, "TEN_GOD_HANJA", { enumerable: true, get: function () { return constants_2.TEN_GOD_HANJA; } });
var elements_2 = require("./elements");
Object.defineProperty(exports, "getHeavenlyStemYinYang", { enumerable: true, get: function () { return elements_2.getHeavenlyStemYinYang; } });
Object.defineProperty(exports, "getHeavenlyStemElement", { enumerable: true, get: function () { return elements_2.getHeavenlyStemElement; } });
Object.defineProperty(exports, "getEarthlyBranchYinYang", { enumerable: true, get: function () { return elements_2.getEarthlyBranchYinYang; } });
Object.defineProperty(exports, "getEarthlyBranchElement", { enumerable: true, get: function () { return elements_2.getEarthlyBranchElement; } });
var convert_2 = require("./calendar/convert");
Object.defineProperty(exports, "lunarToSolar", { enumerable: true, get: function () { return convert_2.lunarToSolar; } });
Object.defineProperty(exports, "solarToLunar", { enumerable: true, get: function () { return convert_2.solarToLunar; } });
Object.defineProperty(exports, "isValidSolarDate", { enumerable: true, get: function () { return convert_2.isValidSolarDate; } });
var lunar_data_2 = require("./calendar/lunar-data");
Object.defineProperty(exports, "LUNAR_MIN_YEAR", { enumerable: true, get: function () { return lunar_data_2.LUNAR_MIN_YEAR; } });
Object.defineProperty(exports, "LUNAR_MAX_YEAR", { enumerable: true, get: function () { return lunar_data_2.LUNAR_MAX_YEAR; } });
var solar_terms_1 = require("./astro/solar-terms");
Object.defineProperty(exports, "getSolarTerm", { enumerable: true, get: function () { return solar_terms_1.getSolarTerm; } });
Object.defineProperty(exports, "getSolarTermsOfYear", { enumerable: true, get: function () { return solar_terms_1.getSolarTermsOfYear; } });
Object.defineProperty(exports, "SOLAR_TERM_NAMES", { enumerable: true, get: function () { return solar_terms_1.SOLAR_TERM_NAMES; } });
Object.defineProperty(exports, "SOLAR_TERM_NAMES_HANJA", { enumerable: true, get: function () { return solar_terms_1.SOLAR_TERM_NAMES_HANJA; } });
var sun_longitude_1 = require("./astro/sun-longitude");
Object.defineProperty(exports, "apparentSolarLongitude", { enumerable: true, get: function () { return sun_longitude_1.apparentSolarLongitude; } });
Object.defineProperty(exports, "equationOfTimeMinutes", { enumerable: true, get: function () { return sun_longitude_1.equationOfTimeMinutes; } });
var ten_gods_2 = require("./features/ten-gods");
Object.defineProperty(exports, "getTenGod", { enumerable: true, get: function () { return ten_gods_2.getTenGod; } });
Object.defineProperty(exports, "getBranchTenGod", { enumerable: true, get: function () { return ten_gods_2.getBranchTenGod; } });
Object.defineProperty(exports, "getTenGodChart", { enumerable: true, get: function () { return ten_gods_2.getTenGodChart; } });
var void_branches_2 = require("./features/void-branches");
Object.defineProperty(exports, "getVoidBranches", { enumerable: true, get: function () { return void_branches_2.getVoidBranches; } });
var luck_pillars_2 = require("./features/luck-pillars");
Object.defineProperty(exports, "getLuckPillars", { enumerable: true, get: function () { return luck_pillars_2.getLuckPillars; } });
var true_solar_time_2 = require("./time/true-solar-time");
Object.defineProperty(exports, "DEFAULT_LONGITUDE", { enumerable: true, get: function () { return true_solar_time_2.DEFAULT_LONGITUDE; } });
// ===== 입력 검증 =====
function validateBirthInfo(birthInfo) {
    if (birthInfo === null || typeof birthInfo !== 'object') {
        throw new TypeError('생년월일시 정보(birthInfo)는 객체여야 합니다.');
    }
    const { year, month, day, hour, minute } = birthInfo;
    if (birthInfo.isLunar !== undefined) {
        (0, validation_1.assertBoolean)(birthInfo.isLunar, '음력 여부(isLunar)');
    }
    if (birthInfo.isLeapMonth !== undefined) {
        (0, validation_1.assertBoolean)(birthInfo.isLeapMonth, '윤달 여부(isLeapMonth)');
    }
    if (birthInfo.dayBoundary !== undefined) {
        (0, validation_1.assertDayBoundary)(birthInfo.dayBoundary);
    }
    if (birthInfo.gender !== undefined) {
        (0, validation_1.assertGender)(birthInfo.gender);
    }
    if (birthInfo.trueSolarTime !== undefined) {
        const { trueSolarTime } = birthInfo;
        if (trueSolarTime === null || typeof trueSolarTime !== 'object' || Array.isArray(trueSolarTime)) {
            throw new TypeError('진태양시 옵션(trueSolarTime)은 객체여야 합니다.');
        }
        if (trueSolarTime.longitude !== undefined) {
            (0, validation_1.assertFiniteNumber)(trueSolarTime.longitude, '출생지 경도(trueSolarTime.longitude)');
            if (trueSolarTime.longitude < -180 || trueSolarTime.longitude > 180) {
                throw new RangeError(`출생지 경도(trueSolarTime.longitude)는 -180~180 범위여야 합니다: ${trueSolarTime.longitude}`);
            }
        }
        (0, validation_1.assertOptionalBoolean)(trueSolarTime.applyEquationOfTime, '균시차 보정 여부(trueSolarTime.applyEquationOfTime)');
        (0, validation_1.assertOptionalBoolean)(trueSolarTime.applyHistoricalDst, '과거 표준시·서머타임 보정 여부(trueSolarTime.applyHistoricalDst)');
    }
    // 사주 계산은 정밀 절입표 범위(1800~) 안에서만 분 단위로 정확하므로 음력 입력도 1800년부터
    // 허용한다. 음력 상한은 음력 테이블(~2100), 양력은 절입표(~2300) 범위를 따른다.
    // (음양력 변환 API 자체는 1391년부터 지원한다.)
    const minYear = solar_terms_data_1.SOLAR_TERM_DATA_MIN_YEAR;
    const maxYear = birthInfo.isLunar ? lunar_data_1.LUNAR_MAX_YEAR : solar_terms_data_1.SOLAR_TERM_DATA_MAX_YEAR;
    if (!Number.isInteger(year) || year < minYear || year > maxYear) {
        throw new RangeError(`연도(year)는 ${minYear}~${maxYear} 정수여야 합니다: ${year}`);
    }
    if (!Number.isInteger(hour) || hour < 0 || hour > 23) {
        throw new RangeError(`시(hour)는 0~23 정수여야 합니다: ${hour}`);
    }
    if (!Number.isInteger(minute) || minute < 0 || minute > 59) {
        throw new RangeError(`분(minute)은 0~59 정수여야 합니다: ${minute}`);
    }
    if (birthInfo.isLunar) {
        // 음력은 월 1~12, 일 1~30 기본 검사. 윤달·일수 세부 검증은 lunarToSolar 가 수행한다.
        if (!Number.isInteger(month) || month < 1 || month > 12) {
            throw new RangeError(`월(month)은 1~12 정수여야 합니다: ${month}`);
        }
        if (!Number.isInteger(day) || day < 1 || day > 30) {
            throw new RangeError(`음력 일(day)은 1~30 정수여야 합니다: ${day}`);
        }
    }
    else if (!(0, convert_1.isValidSolarDate)(year, month, day)) {
        throw new RangeError(`유효하지 않은 양력 날짜입니다: ${year}-${month}-${day}`);
    }
}
function hanjaOf(pillar) {
    return (constants_1.HEAVENLY_STEMS_HANJA[constants_1.HEAVENLY_STEMS.indexOf(pillar.heavenlyStem)] +
        constants_1.EARTHLY_BRANCHES_HANJA[constants_1.EARTHLY_BRANCHES.indexOf(pillar.earthlyBranch)]);
}
function elementOf(pillar) {
    return {
        stem: (0, elements_1.getHeavenlyStemElement)(pillar.heavenlyStem),
        branch: (0, elements_1.getEarthlyBranchElement)(pillar.earthlyBranch),
    };
}
function yinYangOf(pillar) {
    return {
        stem: (0, elements_1.getHeavenlyStemYinYang)(pillar.heavenlyStem),
        branch: (0, elements_1.getEarthlyBranchYinYang)(pillar.earthlyBranch),
    };
}
/**
 * 사주팔자를 계산합니다.
 *
 * @param birthInfo 생년월일시 정보
 * @returns 사주팔자(연주·월주·일주·시주)와 부가 명리 정보
 */
function calculateFourPillars(birthInfo) {
    validateBirthInfo(birthInfo);
    const { hour, minute } = birthInfo;
    let { year, month, day } = birthInfo;
    // 음력 입력 → 양력 변환
    if (birthInfo.isLunar) {
        const solar = (0, convert_1.lunarToSolar)(year, month, day, birthInfo.isLeapMonth ?? false);
        year = solar.year;
        month = solar.month;
        day = solar.day;
    }
    const resolved = (0, true_solar_time_1.resolveInstant)(year, month, day, hour, minute, birthInfo.trueSolarTime);
    const dayBoundary = birthInfo.dayBoundary ?? 'midnight';
    const pillars = (0, pillars_1.computeFourPillars)(resolved, year, dayBoundary);
    const fourPillars = {
        year: pillars.year,
        month: pillars.month,
        day: pillars.day,
        hour: pillars.hour,
    };
    // 부가 명리 정보
    const tenGods = (0, ten_gods_1.getTenGodChart)(fourPillars);
    const voidBranches = (0, void_branches_1.getVoidBranches)(pillars.day.heavenlyStem, pillars.day.earthlyBranch);
    let luckPillars;
    if (birthInfo.gender) {
        luckPillars = (0, luck_pillars_1.getLuckPillars)({
            instantUTCms: resolved.instantUTCms,
            birthYear: year,
            monthPillar: pillars.month,
            sajuYearStemIndex: constants_1.HEAVENLY_STEMS.indexOf(pillars.year.heavenlyStem),
            gender: birthInfo.gender,
        });
    }
    const yearString = `${pillars.year.heavenlyStem}${pillars.year.earthlyBranch}`;
    const monthString = `${pillars.month.heavenlyStem}${pillars.month.earthlyBranch}`;
    const dayString = `${pillars.day.heavenlyStem}${pillars.day.earthlyBranch}`;
    const hourString = `${pillars.hour.heavenlyStem}${pillars.hour.earthlyBranch}`;
    const yearHanja = hanjaOf(pillars.year);
    const monthHanja = hanjaOf(pillars.month);
    const dayHanja = hanjaOf(pillars.day);
    const hourHanja = hanjaOf(pillars.hour);
    return {
        ...fourPillars,
        yearElement: elementOf(pillars.year),
        monthElement: elementOf(pillars.month),
        dayElement: elementOf(pillars.day),
        hourElement: elementOf(pillars.hour),
        yearYinYang: yinYangOf(pillars.year),
        monthYinYang: yinYangOf(pillars.month),
        dayYinYang: yinYangOf(pillars.day),
        hourYinYang: yinYangOf(pillars.hour),
        yearString,
        monthString,
        dayString,
        hourString,
        yearHanja,
        monthHanja,
        dayHanja,
        hourHanja,
        tenGods,
        voidBranches,
        luckPillars,
        toString() {
            return fourPillarsToString(fourPillars);
        },
        toObject() {
            return { year: yearString, month: monthString, day: dayString, hour: hourString };
        },
        toHanjaObject() {
            return {
                year: { korean: yearString, hanja: yearHanja },
                month: { korean: monthString, hanja: monthHanja },
                day: { korean: dayString, hanja: dayHanja },
                hour: { korean: hourString, hanja: hourHanja },
            };
        },
        toHanjaString() {
            return `${yearHanja}年柱, ${monthHanja}月柱, ${dayHanja}日柱, ${hourHanja}時柱`;
        },
    };
}
/**
 * 사주를 한국어 문자열로 변환합니다.
 *
 * @returns "임신연주, 경술월주, 계유일주, 을묘시주" 형식의 문자열
 */
function fourPillarsToString(fourPillars) {
    const { year, month, day, hour } = fourPillars;
    (0, validation_1.assertPillar)(year, 'year');
    (0, validation_1.assertPillar)(month, 'month');
    (0, validation_1.assertPillar)(day, 'day');
    (0, validation_1.assertPillar)(hour, 'hour');
    return [
        `${year.heavenlyStem}${year.earthlyBranch}연주`,
        `${month.heavenlyStem}${month.earthlyBranch}월주`,
        `${day.heavenlyStem}${day.earthlyBranch}일주`,
        `${hour.heavenlyStem}${hour.earthlyBranch}시주`,
    ].join(', ');
}
