// manseryeok/dist/validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertIntegerInRange = assertIntegerInRange;
exports.assertFiniteNumber = assertFiniteNumber;
exports.assertBoolean = assertBoolean;
exports.assertOptionalBoolean = assertOptionalBoolean;
exports.assertHeavenlyStem = assertHeavenlyStem;
exports.assertEarthlyBranch = assertEarthlyBranch;
exports.assertPillar = assertPillar;
exports.assertGender = assertGender;
exports.assertDayBoundary = assertDayBoundary;
const constants_1 = require("./constants");
function assertIntegerInRange(value, min, max, name) {
    if (typeof value !== 'number' || !Number.isInteger(value)) {
        throw new RangeError(`${name}은 정수여야 합니다: ${String(value)}`);
    }
    if (value < min || value > max) {
        throw new RangeError(`${name}은 ${min}~${max} 범위여야 합니다: ${value}`);
    }
}
function assertFiniteNumber(value, name) {
    if (typeof value !== 'number' || !Number.isFinite(value)) {
        throw new RangeError(`${name}은 유한한 숫자여야 합니다: ${String(value)}`);
    }
}
function assertBoolean(value, name) {
    if (typeof value !== 'boolean') {
        throw new TypeError(`${name}은 boolean 이어야 합니다: ${String(value)}`);
    }
}
function assertOptionalBoolean(value, name) {
    if (value !== undefined) {
        assertBoolean(value, name);
    }
}
function assertHeavenlyStem(value, name = '천간') {
    if (typeof value !== 'string' || !constants_1.HEAVENLY_STEMS.includes(value)) {
        throw new RangeError(`${name}은 유효한 천간이어야 합니다: ${String(value)}`);
    }
}
function assertEarthlyBranch(value, name = '지지') {
    if (typeof value !== 'string' || !constants_1.EARTHLY_BRANCHES.includes(value)) {
        throw new RangeError(`${name}은 유효한 지지여야 합니다: ${String(value)}`);
    }
}
function assertPillar(value, name = '기둥') {
    if (value === null || typeof value !== 'object') {
        throw new TypeError(`${name}은 객체여야 합니다.`);
    }
    const pillar = value;
    assertHeavenlyStem(pillar.heavenlyStem, `${name}.heavenlyStem`);
    assertEarthlyBranch(pillar.earthlyBranch, `${name}.earthlyBranch`);
}
function assertGender(value, name = '성별(gender)') {
    if (value !== 'male' && value !== 'female') {
        throw new RangeError(`${name}은 'male' 또는 'female' 이어야 합니다: ${String(value)}`);
    }
}
function assertDayBoundary(value, name = '일 경계(dayBoundary)') {
    if (value !== 'midnight' && value !== 'jasi' && value !== 'splitJasi') {
        throw new RangeError(`${name}는 'midnight', 'jasi', 'splitJasi' 중 하나여야 합니다: ${String(value)}`);
    }
}
