// manseryeok/dist/features/ten-gods.js
"use strict";
/**
 * 십신(十神) 계산.
 *
 * 일간(日干)을 기준으로 다른 천간/지지(지장간 본기)와의 오행 생극 + 음양 관계로
 * 비견·겁재·식신·상관·편재·정재·편관·정관·편인·정인을 산출한다.
 *
 * 규칙: 같은 음양이면 편(偏)계열(비견·식신·편재·편관·편인),
 *       다른 음양이면 정(正)계열(겁재·상관·정재·정관·정인).
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTenGod = getTenGod;
exports.getBranchTenGod = getBranchTenGod;
exports.getTenGodChart = getTenGodChart;
const constants_1 = require("../constants");
const elements_1 = require("../elements");
const validation_1 = require("../validation");
/**
 * 일간 기준 어떤 천간의 십신을 구한다.
 */
function getTenGod(dayMaster, target) {
    const dayEl = (0, elements_1.getHeavenlyStemElement)(dayMaster);
    const targetEl = (0, elements_1.getHeavenlyStemElement)(target);
    const sameYinYang = (0, elements_1.getHeavenlyStemYinYang)(dayMaster) === (0, elements_1.getHeavenlyStemYinYang)(target);
    if (targetEl === dayEl) {
        return sameYinYang ? '비견' : '겁재';
    }
    if (constants_1.ELEMENT_GENERATES[dayEl] === targetEl) {
        return sameYinYang ? '식신' : '상관';
    }
    if (constants_1.ELEMENT_CONTROLS[dayEl] === targetEl) {
        return sameYinYang ? '편재' : '정재';
    }
    if (constants_1.ELEMENT_CONTROLS[targetEl] === dayEl) {
        return sameYinYang ? '편관' : '정관';
    }
    // 나머지: target 이 일간을 생함 (인성)
    return sameYinYang ? '편인' : '정인';
}
/**
 * 일간 기준 어떤 지지의 십신을 구한다 (지장간 본기 기준).
 */
function getBranchTenGod(dayMaster, branch) {
    (0, validation_1.assertEarthlyBranch)(branch);
    return getTenGod(dayMaster, constants_1.BRANCH_MAIN_STEM[branch]);
}
/**
 * 사주팔자 전체의 십신 차트를 계산한다.
 */
function getTenGodChart(pillars) {
    const dayMaster = pillars.day.heavenlyStem;
    return {
        year: {
            stem: getTenGod(dayMaster, pillars.year.heavenlyStem),
            branch: getBranchTenGod(dayMaster, pillars.year.earthlyBranch),
        },
        month: {
            stem: getTenGod(dayMaster, pillars.month.heavenlyStem),
            branch: getBranchTenGod(dayMaster, pillars.month.earthlyBranch),
        },
        day: {
            stem: '일간',
            branch: getBranchTenGod(dayMaster, pillars.day.earthlyBranch),
        },
        hour: {
            stem: getTenGod(dayMaster, pillars.hour.heavenlyStem),
            branch: getBranchTenGod(dayMaster, pillars.hour.earthlyBranch),
        },
    };
}
